Use target model's `destroy_association_async_job` when available

hubb opened the pull request:
### Motivation / Background

When destroying associations with `dependent: :destroy_async`, Rails now respects the target model's `destroy_association_async_job` configuration instead of always using the owner model's job.

This allows models to define their own custom destruction jobs that will be used regardless of which association triggered the destruction:

```
    class Parent < ApplicationRecord
      self.destroy_association_async_job = ParentDestroyJob
      has_many :children, dependent: :destroy_async
    end

    class Child < ApplicationRecord
      self.destroy_association_async_job = ChildDestroyJob
    end
```

Previously, destroying a parent would use `ParentDestroyJob` for the children. Now it correctly uses `ChildDestroyJob`.

The change maintains backward compatibility by falling back to the owner model's job when the target model has no custom job configured.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

hubb committed:
Use target model's destroy_association_async_job when available

When destroying associations with `dependent: :destroy_async`, Rails now
respects the target model's `destroy_association_async_job` configuration
instead of always using the owner model's job.

This allows models to define their own custom destruction jobs that will
be used regardless of which association triggered the destruction:

    class Parent < ApplicationRecord
      self.destroy_association_async_job = ParentDestroyJob
      has_many :children, dependent: :destroy_async
    end

    class Child < ApplicationRecord
      self.destroy_association_async_job = ChildDestroyJob
    end

Previously, destroying a parent would use ParentDestroyJob for the
children. Now it correctly uses ChildDestroyJob.

The change maintains backward compatibility by falling back to the
owner model's job when the target model has no custom job configured.

rafaelfranca reviewed (changes requested):
Not sure this new behavior makes sense. I think `destroy_association_async_job` was made to change the job of all associations declared in the model. If we want to change which job will run to destroy a single association we should pass it as option to the association.

```
has_many :children, dependent: :destroy_async, destroy_job: "SomeJob"`

rafaelfranca reviewed (commented):
I think the behavior you want makes more sense. It was never the intention to use the job in the owner class. I think we can just swap this behavior using a configuration option to allow people to move to the new behavior.

matthewd commented:
> It was never the intention to use the job in the owner class

@rafaelfranca I disagree; it's a configuration on the owner class, describing how to handle its association. The arguments to the job are the owner instance and the association name -- that very much feels like something that belongs to the owner, not the association target class (without even contemplating the possibilities of STI).

The behaviour proposed here would be a `destroy_async_job` or something, that described how to destroy the model being configured.

The current behaviour seems like it would be desirable for anyone implementing e.g. a papertrail-type versioning system for a given model and its associations.

hubb committed:
Revert "Use target model's destroy_association_async_job when available"

This reverts commit fc5113218f73805d6c5952f319494a8fa202a7e2.

hubb committed:
Add destroy_job option to associations for customizing async destruction

Adds a new 'destroy_job' option to association definitions that allows specifying
a custom job class to use when destroying associations with dependent: :destroy_async.

Example:
  has_many :children, dependent: :destroy_async, destroy_job: CustomDestroyJob
  has_one :profile, dependent: :destroy_async, destroy_job: 'ProfileCleanupJob'
  belongs_to :org, dependent: :destroy_async, destroy_job: OrgDestroyJob

This provides fine-grained control over which job processes the destruction of
associated records, taking precedence over the model's destroy_association_async_job
configuration when specified.
